src/dbnode/client/session_test.go

```diff
@@ -445,25 +445,31 @@ func TestDedicatedConnection(t *testing.T) {
 
 	_, ch, err := s.DedicatedConnection(shardID, DedicatedConnectionOptions{})
 	require.NoError(t, err)
-	assert.Equal(t, &noopPooledChannel{"remote1"}, ch)
+	assert.Equal(t, "remote1", asNoopPooledChannel(ch).address)
 
 	_, ch2, err := s.DedicatedConnection(shardID, DedicatedConnectionOptions{ShardStateFilter: shard.Available})
 	require.NoError(t, err)
-	assert.Equal(t, &noopPooledChannel{"remote2"}, ch2)
+	assert.Equal(t, "remote2", asNoopPooledChannel(ch2).address)
 
 	s.healthCheckNewConnFn = testHealthCheck(nil, true)
 	_, ch3, err := s.DedicatedConnection(shardID, DedicatedConnectionOptions{BootstrappedNodesOnly: true})
 	require.NoError(t, err)
-	assert.Equal(t, &noopPooledChannel{"remote1"}, ch3)
+	assert.Equal(t, "remote1", asNoopPooledChannel(ch3).address)
 
 	healthErr := errors.New("unhealthy")
 	s.healthCheckNewConnFn = testHealthCheck(healthErr, false)
 
+	ch = &noopPooledChannel{"test", 0}
+	s.opts = NewOptions().SetNewConnectionFn(func(_ string, _ string, _ Options) (Channel, rpc.TChanNode, error) {
+		return ch, nil, nil
+	})
 	_, _, err = s.DedicatedConnection(shardID, DedicatedConnectionOptions{})
 	require.NotNil(t, err)
 	multiErr, ok := err.(xerror.MultiError) // nolint: errorlint
 	assert.True(t, ok, "expecting MultiError")
 	assert.True(t, multiErr.Contains(healthErr))
+	// 2 because of 2 remote hosts failing health check
+	assert.Equal(t, 2, asNoopPooledChannel(ch).CloseCount())
```

Comment:
I think this is not correct. There should be no double close (and if there is, the test should fail). 
This happens in test code because you return the same `ch` from the pool twice, which does not model the actual production code behaviour accurately.

Reply:
You could simply have a slice of chs to which you would append from `SetNewConnectionFn(func() {...})` and then in the end would assert that the slice contains 2 chs and that each of them has been closed exactly once.